go/vt/vtgate/vindexes/vschema.go

```diff
@@ -245,6 +248,116 @@ func buildKeyspaces(source *vschemapb.SrvVSchema, vschema *VSchema) {
 		vschema.Keyspaces[ksname] = ksvschema
 		ksvschema.Error = buildTables(ks, vschema, ksvschema)
 	}
+
+	// In the second loop, build references.
+	for ksname, ks := range source.Keyspaces {
+		ksvschema := vschema.Keyspaces[ksname]
+		if err := buildReferences(ks, vschema, ksvschema, allowUnresolvedReferences); err != nil && ksvschema.Error == nil {
+			ksvschema.Error = err
+		}
+	}
+
+	// In the third loop, build global tables.
+	for ksname, ks := range source.Keyspaces {
+		ksvschema := vschema.Keyspaces[ksname]
+		buildGlobalTables(ks, vschema, ksvschema)
+	}
+}
+
+func buildGlobalTables(ks *vschemapb.Keyspace, vschema *VSchema, ksvschema *KeyspaceSchema) {
+	for tname, t := range ksvschema.Tables {
+		// If the keyspace requires explicit routing, don't include any of
+		// its tables in global tables.
+		if !ks.RequireExplicitRouting {
+			if gt, ok := vschema.globalTables[tname]; ok {
+				// There is already an entry table stored in global tables
+				// with this name.
+				if gt == nil {
+					// Table name is already marked ambiguous, nothing to do.
+					continue
+				} else if _, ok := t.ReferencedBy[gt.Keyspace.Name]; ok {
+					// If the stored table refers to this table, store this
+					// table instead.
+					vschema.globalTables[tname] = t
+				} else if _, ok := gt.ReferencedBy[t.Keyspace.Name]; ok {
```

Comment:
these can throw `map nil` error as they are not always initialized.
nil check would be needed here.